File: fsharp-webgl/Client/WebGL2RenderingContext.fs
namespace Client

open Microsoft.JSInterop
open System

type WebGL2RenderingContext(js: IJSInProcessRuntime, context: IJSUnmarshalledObjectReference) =
    member val COLOR_BUFFER_BIT = js.Invoke<int>("getValue", context, "COLOR_BUFFER_BIT")

    member val VERTEX_SHADER = js.Invoke<int>("getValue", context, "VERTEX_SHADER")
    member val FRAGMENT_SHADER = js.Invoke<int>("getValue", context, "FRAGMENT_SHADER")

    member val COMPILE_STATUS = js.Invoke<int>("getValue", context, "COMPILE_STATUS")
    member val LINK_STATUS = js.Invoke<int>("getValue", context, "LINK_STATUS")
    member val ACTIVE_ATTRIBUTES = js.Invoke<int>("getValue", context, "ACTIVE_ATTRIBUTES")
    member val ACTIVE_UNIFORMS = js.Invoke<int>("getValue", context, "ACTIVE_UNIFORMS")

    member val ARRAY_BUFFER = js.Invoke<int>("getValue", context, "ARRAY_BUFFER")

    member val STATIC_DRAW = js.Invoke<int>("getValue", context, "STATIC_DRAW")

    member this.arrayToFloat32Array(input: single array) =
        (js :?> IJSUnmarshalledRuntime)
            .InvokeUnmarshalled<float32 array, IJSUnmarshalledObjectReference>("arrayToFloat32Array", input)

    member this.clearColor (r: float) (g: float) (b: float) (a: float) =
        context.InvokeVoid("clearColor", r, g, b, a)

    member this.clear(bits: int) = context.InvokeVoid("clear", bits)

    member this.viewport (x: int) (y: int) (width: int) (height: int) =
        context.InvokeVoid("viewport", x, y, width, height)

    member this.createShader(typ: int) =
        context.Invoke<IJSInProcessObjectReference>("createShader", typ)

    member this.shaderSource (shader: IJSInProcessObjectReference) (source: string) =
        context.InvokeVoid("shaderSource", shader, source)

    member this.compileShader(shader: IJSInProcessObjectReference) =
        context.InvokeVoid("compileShader", shader)

    member this.getShaderParameter<'T> (shader: IJSInProcessObjectReference) (param: int) =
        context.Invoke<'T>("getShaderParameter", shader, param)

    member this.getShaderInfoLog(shader: IJSInProcessObjectReference) =
        context.Invoke<string>("getShaderInfoLog", shader)

    member this.deleteShader(shader: IJSInProcessObjectReference) =
        context.InvokeVoid("deleteShader", shader)

    member this.createProgram() =
        context.Invoke<IJSInProcessObjectReference>("createProgram")

    member this.attachShader (program: IJSInProcessObjectReference) (shader: IJSInProcessObjectReference) =
        context.InvokeVoid("attachShader", program, shader)

    member this.linkProgram(program: IJSInProcessObjectReference) =
        context.InvokeVoid("linkProgram", program)

    member this.getProgramParameter<'T> (program: IJSInProcessObjectReference) (param: int) =
        context.Invoke<'T>("getProgramParameter", program, param)

    member this.getProgramInfoLog(program: IJSInProcessObjectReference) =
        context.Invoke<string>("getProgramInfoLog", program)

    member this.deleteProgram(program: IJSInProcessObjectReference) =
        context.InvokeVoid("deleteProgram", program)

    member this.getActiveAttrib (program: IJSInProcessObjectReference) (index: int) =
        context.Invoke<IJSInProcessObjectReference>("getActiveAttrib", program, index)

    member this.getActiveUniform (program: IJSInProcessObjectReference) (index: int) =
        context.Invoke<IJSInProcessObjectReference>("getActiveUniform", program, index)

    member this.createBuffer() =
        context.Invoke<IJSInProcessObjectReference>("createBuffer")

    member this.bindBuffer (typ: int) (buffer: IJSInProcessObjectReference option) =
        context.InvokeVoid(
            "bindBuffer",
            typ,
            (match buffer with
             | Some(buffer) -> buffer
             | None -> null)
        )

    member this.bufferData (typ: int) (data: IJSUnmarshalledObjectReference) (usage: int) =
        context.InvokeVoid("bufferData", typ, data, usage)

File: csharp-blazor-webgl/BlazorExperiments/WebGl/WebGL2RenderingContext.cs
using Microsoft.JSInterop;

namespace BlazorExperiments.WebGl;

public class WebGL2RenderingContext
{
    public record class Shader(IJSInProcessObjectReference ObjRef);
    public record class ShaderProgram(IJSInProcessObjectReference ObjRef);
    public record class Buffer(IJSInProcessObjectReference ObjRef);

    public const int COLOR_BUFFER_BIT = 0x00004000;

    public const int FRAGMENT_SHADER = 0x8B30;
    public const int VERTEX_SHADER = 0x8B31;

    public const int COMPILE_STATUS = 0x8B81;
    public const int LINK_STATUS = 0x8B82;

    public const int ARRAY_BUFFER = 0x8892;
    public const int ELEMENT_ARRAY_BUFFER = 0x8893;

    public const int STATIC_DRAW = 0x88E4;
    public const int STREAM_DRAW = 0x88E0;
    public const int DYNAMIC_DRAW = 0x88E8;
    public const int STREAM_READ = 0x88E1;
    public const int STREAM_COPY = 0x88E2;
    public const int STATIC_READ = 0x88E5;
    public const int STATIC_COPY = 0x88E6;
    public const int DYNAMIC_READ = 0x88E9;
    public const int DYNAMIC_COPY = 0x88EA;

    public const int BYTE = 0x1400;
    public const int UNSIGNED_BYTE = 0x1401;
    public const int SHORT = 0x1402;
    public const int UNSIGNED_SHORT = 0x1403;
    public const int INT = 0x1404;
    public const int UNSIGNED_INT = 0x1405;
    public const int FLOAT = 0x1406;

    public const int POINTS = 0x0000;
    public const int LINES = 0x0001;
    public const int LINE_LOOP = 0x0002;
    public const int LINE_STRIP = 0x0003;
    public const int TRIANGLES = 0x0004;
    public const int TRIANGLE_STRIP = 0x0005;
    public const int TRIANGLE_FAN = 0x0006;

    private readonly IJSInProcessObjectReference objRef;

    public WebGL2RenderingContext(IJSInProcessObjectReference objRef)
    {
        this.objRef = objRef;
    }

    public void Viewport(int x, int y, int width, int height)
    {
        objRef.InvokeVoid("viewport", x, y, width, height);
    }

    public void ClearColor(double red, double green, double blue, double alpha)
    {
        objRef.InvokeVoid("clearColor", red, green, blue, alpha);
    }

    public void Clear(int bits)
    {
        objRef.InvokeVoid("clear", bits);
    }

    public Shader CreateShader(int type)
    {
        return new Shader(objRef.Invoke<IJSInProcessObjectReference>("createShader", type));
    }

    public void DeleteShader(Shader shader)
    {
        objRef.InvokeVoid("deleteShader", shader.ObjRef);
    }

    public void ShaderSource(Shader shader, string source)
    {
        objRef.InvokeVoid("shaderSource", shader.ObjRef, source);
    }

    public void CompileShader(Shader shader)
    {
        objRef.InvokeVoid("compileShader", shader.ObjRef);
    }

    public T GetShaderParameter<T>(Shader shader, int pname)
    {
        return objRef.Invoke<T>("getShaderParameter", shader.ObjRef, pname);
    }

    public string GetShaderInfoLog(Shader program)
    {
        return objRef.Invoke<string>("getShaderInfoLog", program.ObjRef);
    }

    public ShaderProgram CreateProgram()
    {
        return new ShaderProgram(objRef.Invoke<IJSInProcessObjectReference>("createProgram"));
    }

    public void DeleteProgram(ShaderProgram program)
    {
        objRef.InvokeVoid("deleteProgram", program.ObjRef);
    }

    public void AttachShader(ShaderProgram program, Shader shader)
    {
        objRef.InvokeVoid("attachShader", program.ObjRef, shader.ObjRef);
    }

    public void DetachShader(ShaderProgram program, Shader shader)
    {
        objRef.InvokeVoid("detachShader", program.ObjRef, shader.ObjRef);
    }

    public void LinkProgram(ShaderProgram program)
    {
        objRef.InvokeVoid("linkProgram", program.ObjRef);
    }

    public T GetProgramParameter<T>(ShaderProgram program, int pname)
    {
        return objRef.Invoke<T>("getProgramParameter", program.ObjRef, pname);
    }

    public string GetProgramInfoLog(ShaderProgram program)
    {
        return objRef.Invoke<string>("getProgramInfoLog", program.ObjRef);
    }

    public int GetAttribLocation(ShaderProgram program, string name)
    {
        return objRef.Invoke<int>("getAttribLocation", program.ObjRef, name);
    }

    // TODO various shader methods for getting attributes and uniforms

    public void UseProgram(ShaderProgram? program)
    {
        objRef.InvokeVoid("useProgram", program?.ObjRef);
    }

    public Buffer CreateBuffer()
    {
        return new Buffer(objRef.Invoke<IJSInProcessObjectReference>("createBuffer"));
    }

    public void DeleteBuffer(Buffer buffer)
    {
        objRef.InvokeVoid("deleteBuffer", buffer.ObjRef);
    }

    public void BindBuffer(int type, Buffer? buffer)
    {
        objRef.InvokeVoid("bindBuffer", type, buffer?.ObjRef);
    }

    // TODO BufferData with size
    // TODO BufferData with byte[]

    public void BufferData(int type, float[] data, int usage)
    {
        objRef.InvokeVoid("bufferData_float32", type, data, usage);
    }

    // TODO BufferData with double[]

    public void EnableVertexAttribArray(int i)
    {
        objRef.InvokeVoid("enableVertexAttribArray", i);
    }

    public void DisableVertexAttribArray(int i)
    {
        objRef.InvokeVoid("disableVertexAttribArray", i);
    }

    public void VertexAttribPointer(int index, int size, int type, bool normalized, int stride, int offset)
    {
        objRef.InvokeVoid("vertexAttribPointer", index, size, type, normalized, stride, offset);
    }

    public void DrawArrays(int mode, int first, int count)
    {
        objRef.InvokeVoid("drawArrays", mode, first, count);
    }

    // TODO drawElements
}
